proof_of_stake/src/epoched.rs

```diff
@@ -0,0 +1,1270 @@
+        current_epoch: impl Into<Epoch>,
+        params: &PosParams,
+    ) {
+        let epoch = current_epoch.into();
+        debug_assert!(
+            epoch >= self.last_update,
+            "The current epoch must be greater than or equal to the last \
+             update"
+        );
+        let offset = Offset::value(params) as usize;
+        let last_update = self.last_update;
+        let shift: usize =
+            cmp::min((epoch.sub_or_default(last_update)).into(), offset);
+
+        // Resize the data if needed
+        if self.data.len() < offset + 1 {
+            self.data.resize_with(offset + 1, Default::default);
+        }
+
+        if shift != 0 {
+            let mid_point = cmp::min(shift, self.data.len());
+            let mut latest_value: Option<Data> = None;
+            // Find the latest value in elements before the mid-point and clear
+            // them
+            for i in 0..mid_point {
+                if let Some(Some(data)) = self.data.get(i) {
+                    latest_value = Some(data.clone());
+                }
+                self.data[i] = None;
+            }
+            // Rotate left on the mid-point
+            self.data.rotate_left(mid_point);
+            // Update the head with the latest value, if it's not already set
+            let current = self.data.get_mut(0).unwrap();
+            if current.is_none() {
+                *current = latest_value;
+            }
+        }
+
+        self.last_update = epoch;
+    }
+
+    /// Set the value at the data's epoch offset.
+    pub fn set(
+        &mut self,
+        value: Data,
+        current_epoch: impl Into<Epoch>,
+        params: &PosParams,
+    ) {
+        self.update_data(current_epoch, params);
+
+        let offset = Offset::value(params) as usize;
+        self.data[offset] = Some(value);
+    }
+
+    /// Update the values starting from the given epoch offset (which must not
+    /// be greater than the `Offset` type parameter of self) with the given
+    /// function.
+    pub fn update_from_offset(
+        &mut self,
+        update_value: impl Fn(&mut Data, Epoch),
+        current_epoch: impl Into<Epoch>,
+        offset: DynEpochOffset,
+        params: &PosParams,
+    ) {
+        let offset = offset.value(params) as usize;
+        debug_assert!(offset <= Offset::value(params) as usize);
+        let epoch = current_epoch.into();
+        self.update_data(epoch, params);
+
+        if let Some(data) = self.data.get_mut(offset).unwrap() {
+            // If there's a value at `offset`, update it
+            update_value(data, self.last_update + offset)
+        } else {
+            // Try to find if there's any value before `offset`
+            let mut latest_value: Option<Data> = None;
+            for i in (0..offset).rev() {
+                if let Some(Some(data)) = self.data.get(i) {
+                    latest_value = Some(data.clone());
+                    break;
+                }
+            }
```

Comment:
```suggestion
            let mut latest_value: Option<&Data> = None;
            for i in (0..offset).rev() {
                if let Some(Some(data)) = self.data.get(i) {
                    latest_value = Some(data);
                    break;
                }
            }
            let latest_value = latest_value.map(|data| data.clone());
```

Reply:
I think in here it's the same thing, as we break the loop after cloning, but I'll commit this anyway with a simplification on the last line:
```suggestion
            let mut latest_value: Option<&Data> = None;
            for i in (0..offset).rev() {
                if let Some(Some(data)) = self.data.get(i) {
                    latest_value = Some(data);
                    break;
                }
            }
            let latest_value = latest_value.cloned();
```